### ISSUE TITLE
Yolov8n trained on custom dataset is not able to detect objects.

### ISSUE THREAD

**Issue by danishali6421** [ultralytics/ultralytics#4969]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

HiTeam,

I have trained Yolov8n model using pretrained model on the custom dataset of football image dataset.  I have trained the model for 10 epochs but the model is not detecting anything even I used same images that I used in the training.  I have implemented the model in local environment using Pycharm.

Kindly suggest some solution.

I used dataset from the kaggle. Link of the dataset is given below,

https://www.kaggle.com/datasets/outliersloop/footballyolov8

The following code was used in training.

from ultralytics import YOLO

# Load a model
#model = YOLO("yolov8n.yaml")  # build a new model from scratch
#model = YOLO(  bn
# Use the model
# Load a pretrained YOLO model (recommended for training)
model = YOLO('yolov8n.pt')

# Train the model using the 'coco128.yaml' dataset for 3 epochs
#results = model.train(data='coco128.yaml', epochs=3)
results=model.train(data="config.yaml", epochs=10)  # train the model
#metrics = model.val()  # evaluate model performance on the validation set
#results = model("https://ultralytics.com/images/bus.jpg")  # predict on an image
#path = model.export(format="onnx")  # export the model to ONNX format

The following code was used for inference:

from ultralytics import YOLO
from PIL import Image
import os
import cv2
model_path = os.path.join('.', 'runs', 'detect', 'train17', 'weights', 'best.pt')
print()
# Load a model
model = YOLO(model_path)  # load a custom model
#model = YOLO("model.pt")
# accepts all formats - image/dir/Path/URL/video/PIL/ndarray. 0 for webcam
#results = model.predict(source="0")
#results = model.predict(source="folder", show=True) # Display preds. Accepts all YOLO predict arguments

# from PIL
im1 = Image.open("bus.jpg")
results = model.predict(source=im1, save=True)  # save plotted images

# from ndarray
#im2 = cv2.imread("bus.jpg")
#results = model.predict(source=im2, save=True, save_txt=True)  # save predictions as labels

# from list of PIL/ndarray
#results = model.predict(source=[im1, im2])

Following results are obtained once training completed.

I am looking forward for your kind support.

**Comment by danishali6421** [comment]
> 👋 Hello @danishali6421, thank you for your interest in YOLOv8 🚀! We recommend a visit to the [YOLOv8 Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.
> 
> If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it.
> 
> If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/).
> 
> Join the vibrant [Ultralytics Discord](https://ultralytics.com/discord) 🎧 community for real-time conversations and collaborations. This platform offers a perfect space to inquire, showcase your work, and connect with fellow Ultralytics users.
> 
> ## Install
> Pip install the `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/requirements.txt) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/).
> 
> ```shell
> pip install ultralytics
> ```
> 
> ## Environments
> YOLOv8 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
> 
> * **Notebooks** with free GPU: [![Run on Gradient](https://camo.githubusercontent.com/c9931a1689c37ab786edd3e1e5f59b9a6f7d097628c4689ce2432563ef884524/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://console.paperspace.com/github/ultralytics/ultralytics) [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/ultralytics/yolov8)
> * **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
> * **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
> * **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) [![Docker Pulls](https://camo.githubusercontent.com/7401823bed1473d6db1ff60b933d51cd1534c448488f06c446c322227d31d966/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f756c7472616c79746963733f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/ultralytics)
> 
> ## Status
> [![Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg)](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule)
> 
> If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLOv8 [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

I have gone through the documentation and following the procedure as mentioned by issue not resolved.

**Comment by RizwanMunawar** [comment]
@danishali6421, thank you for providing the details. Could you please share the Mean Average Precision (MAP), as well as Precision and Recall values, from the training process?

Additionally, I've noticed that you're utilizing the `bus.jpg` image, and I'm uncertain whether it contains football players. Below, you'll find the sample code for running inference:

```python
from ultralytics import YOLO

# Load a custom YOLOv8n model (In your case, you can use 'best.pt')
model = YOLO('path/to/model.pt')

# Perform inference on the image with the following arguments
model.predict('test.jpg', save=True, imgsz=640, conf=0.5)
```

Please note that training the YOLOv8 model for only 10 epochs may not provide the model with a sufficient understanding of football player characteristics. I recommend training the model for at least 50 epochs for better results. Thank you.

**Comment by danishali6421** [comment]
> @danishali6421, thank you for providing the details. Could you please share the Mean Average Precision (MAP), as well as Precision and Recall values, from the training process?
> 
> Additionally, I've noticed that you're utilizing the `bus.jpg` image, and I'm uncertain whether it contains football players. Below, you'll find the sample code for running inference:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load a custom YOLOv8n model (In your case, you can use 'best.pt')
> model = YOLO('path/to/model.pt')
> 
> # Perform inference on the image with the following arguments
> model.predict('test.jpg', save=True, imgsz=640, conf=0.5)
> ```
> 
> Please note that training the YOLOv8 model for only 10 epochs may not provide the model with a sufficient understanding of football player characteristics. I recommend training the model for at least 50 epochs for better results. Thank you.

@RizwanMunawar, thanks for your kind response. bus.jpg is just a typo. I actually test it on the football image dataset. I have attached the result for the MAP and Precision Recall below. 

Furthermore, I want to ask is there any issue on training the model in local CPU environment using Pycharm, Also kindly suggest What is the best environment to train Yolov8 model i-e local environment, Jupyter Notebook or any other keeping in mind that I am using a simple CPU system.


![confusion_matrix_normalized](https://github.com/ultralytics/ultralytics/assets/124576092/5b68e6b6-ff7b-4cbb-94d0-0d36c644bdc1)
![confusion_matrix](https://github.com/ultralytics/ultralytics/assets/124576092/e4227671-c27a-45de-92ed-b2d29a0fd77c)![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/124576092
![F1_curve](https://github.com/ultralytics/ultralytics/assets/124576092/c19abe5b-0fa8-4b44-8e34-ad89602e90ba)
![P_curve](https://github.com/ultralytics/ultralytics/assets/124576092/c14a0689-279f-49d1-9e7b-a16469004519)
![PR_curve](https://github.com/ultralytics/ultralytics/assets/124576092/ef599953-ea11-4bb0-a241-ba413c7c3ed8)
![R_curve](https://github.com/ultralytics/ultralytics/assets/124576092/4a05748a-4fee-4829-a8e9-d50213e43966)
![results](https://github.com/ultralytics/ultralytics/assets/124576092/f8374936-9894-4f32-9764-c20bb7ae678b)
![train_batch0](https://github.com/ultralytics/ultralytics/assets/124576092/7c3e7254-c70a-49e6-9924-d4bf49229fab)
![train_batch1](https://github.com/ultralytics/ultralytics/assets/124576092/7ba1c646-cdf8-4a36-9ef5-006307821243)
/9e6fabe3-d265-4189-b0c4-f75029202246)
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/124576092/61d565a8-56a6-4cc6-80d8-836018249729)
![val_batch1_labels](https://github.com/ultralytics/ultralytics/assets/124576092/a6b03c0a-876a-414a-8cf2-2549f52ac196)

**Comment by RizwanMunawar** [comment]
@danishali6421, the model's accuracy, precision, and recall are not good. Could you please provide information about the number of training images used?

To enhance accuracy, consider the following options:
- Increase the training epochs to a minimum of 50.
- Increase the dataset with additional annotated images of football and people.

Thanks

**Comment by danishali6421** [comment]
> @danishali6421, the model's accuracy, precision, and recall are not good. Could you please provide information about the number of training images used?
> 
> To enhance accuracy, consider the following options:
> 
> * Increase the training epochs to a minimum of 50.
> * Increase the dataset with additional annotated images of football and people.
> 
> Thanks



> @danishali6421, the model's accuracy, precision, and recall are not good. Could you please provide information about the number of training images used?
> 
> To enhance accuracy, consider the following options:
> 
> * Increase the training epochs to a minimum of 50.
> * Increase the dataset with additional annotated images of football and people.
> 
> Thanks



> @danishali6421, the model's accuracy, precision, and recall are not good. Could you please provide information about the number of training images used?
> 
> To enhance accuracy, consider the following options:
> 
> * Increase the training epochs to a minimum of 50.
> * Increase the dataset with additional annotated images of football and people.
> 
> Thanks

@RizwanMunawar I have used small dataset of 128 images in the training. Furthermore, once I predict the model on pre trained Yolov8n.pt it detects images. I think it should detect when I train using pre train model. Is this not the case?

**Comment by RizwanMunawar** [comment]
@danishali6421 you will need to fine-tune the model on more images. 128 images for football and player detection tasks may lead to wrong results. Thanks

**Comment by nuhman-bt** [comment]
> I have trained the model for 10 epochs but the model is not detecting anything even I used same images that I used in the training. I have implemented the model in local environment using Pycharm.

Hey @danishali6421 , think the issue is with the Epochs, Go for Higher epochs for better Result. Minimum 100 Epochs Recommended . Thanks

**Comment by nuhmanpk** [comment]
> I have trained the model for 10 epochs but the model is not detecting anything even I used same images that I used in the training. I have implemented the model in local environment using Pycharm.

Hey @danishali6421 , think the issue is with the Epochs, Go for Higher epochs for better Result. Minimum 100 Epochs Recommended . Thanks

**Comment by glenn-jocher** [comment]
Hi @danishali6421, there could be several reasons behind your model not being able to detect anything. Given the complexity of the YOLOv8 model, training it for only 10 epochs might not be sufficient for it to learn and decipher the features effectively. I suggest you increase the number of training epochs to a minimum of 100 to see noticeable improvements. 

Also, ensure that your training data is sufficiently diverse and comprehensive, covering a wide range of instances of the object of interest under various conditions. The quality and quantity of training data have a massive impact on how well your model will perform. 

Keep training and experimenting!

**Comment by nuhmanpk** [comment]
@danishali6421 To accelerate your training process, I recommend using the GPU or using Google Collab which provides a free GPU . Also recommend using ultralytics hub for better metrics and to store the model and other data related to training in the cloud. Thanks

**Comment by Vikas29001** [comment]
I encountered the same issue, but when I switched to using a Google Colab notebook, it worked perfectly and successfully detected objects. Initially, I was training on a Jupyter notebook using my own GPU, but it wasn't functioning correctly or detecting anything. You should definitely try using a Google Colab notebook with your custom dataset.

**Comment by glenn-jocher** [comment]
Thanks for sharing your experience! 🌟 Switching to Google Colab can indeed offer better performance due to its access to more powerful GPUs compared to most local setups. This can significantly improve training times and detection capabilities. For anyone facing similar issues, it's worth giving Colab a try to leverage their free GPU resources. Here's a quick snippet to set up YOLOv8 on Colab:

```python
from ultralytics import YOLO

# Load a pretrained model
model = YOLO('yolov8n.pt')

# Train the model
results = model.train(data='your_dataset.yaml', epochs=100)
```

Make sure to adjust the dataset path and epochs according to your needs! Happy training! 🚀
